intermediate_source/nvfuser_intro_tutorial.py

```diff
@@ -0,0 +1,647 @@
+# want to validate the output of nvFuser with the output without
+# nvFuser it would require disabling the random number generation
+# functions. In this example we would simply need to change
+# `dropout_out = F.dropout(bias1_out, dropout_prob, training=True)`
+# to
+# `dropout_out = F.dropout(bias1_out, dropout_prob, training=False)`
+# as the dropout function is the only function in this example that
+# depends on random number generation.
+# 
+# .. figure:: /_static/img/nvfuser_intro/nvfuser_tutorial_1.png
+#
+# Our geomean performance with nvFuser is 1,394 images per second
+# which is a geomean of 1.64x faster than eager mode. We did not
+# include the time that TorchScript and nvFuser take to compile the
+# program and GPU functions. For real end-to-end training the
+# compile time of TorchScript and nvFuser are negligible. For
+# example in this tutorial the combination of TorchScript and
+# nvFuser took around 2.4s in total to compile these high speed
+# GPU functions.
+#
+# nvFuser’s capabilities extend well beyond this initial performance gain.
+#
+
+######################################################################
+# nvFuser & Dynamic Shapes
+# ---------------------
+# It is challenging for Deep Learning Compilers to provide performance
+# gains when the user changes the input sizes of the tensors. However,
+# supporting changing shapes has always been a fundamental design
+# criteria for nvFuser, as processing different-sized input tensors is
+# critical to many applications like Natural Language Processing and
+# Graph Neural Networks.
+#
+# To use nvFuser on inputs that change shape from iteration we
+# generate new input and output gradient tensors and make a few
+# different sizes. Since the last dimension is shared with the
+# parameters and cannot be changed dynamically in LayerNorm, we
+# perturb the first two dimensions of the input and gradient tensors.
+#
+
+SHAPE_COUNT = 20
+dynamic_sizes = deepcopy(input_size)
+
+inputs1: List[torch.Tensor] = []
+inputs2: List[torch.Tensor] = []
+grad_outputs: List[torch.Tensor] = []
+
+
+# Create some random shapes
+for _ in range(SHAPE_COUNT):
+    dynamic_sizes[0] = input_size[0] + random.randrange(-2, 3)
+    dynamic_sizes[1] = input_size[1] + random.randrange(-2, 3)
+    input = torch.randn(*dynamic_sizes, device=device,
+                        dtype=dtype, requires_grad=True)
+    inputs1.append(input)
+    inputs2.append(torch.rand_like(input))
+    grad_outputs.append(torch.rand_like(input))
+
+######################################################################
+# No changes from before are required for running with TorchScript, we
+# simply reuse the previous definition that we wrapped in
+# `torch.jit.script`.
+#
+# We’ll start as usual by performing some warm-up iterations, however
+# we won’t show nvFuser all of the input sizes, we’ll only show one
+# size for the warm-up.
+#
+
+# Perform warm-up iterations
```

Comment:
Do we even need to perform the warm-up iterations here? The scripted definition is already compiled from the static example before, it doesn't need recompilation.

Reply:
Yes, I think so as your device might go back to P8 (sleep/idle) which would add a slight overhead in the fist CUDA op to push it back to P0 (perf. mode).